Fixes: #902 - Feature/media background playback

Brandon-T opened the pull request:
Fixes: #902
Added ability to play videos in the background
Added ability to play videos while the device is locked
Added ability to play videos with a passcode enabled device while locked.
Fixed multiple bad crashes when in the background and device is passcode locked while performing any request:
 - Fetch request causes a CoreData write to happen (reportSiteVisited).. however, it can't write and so it crashes instead due to DataProtection being enabled by default. This happens when the video goes to the "next" one in the playlist while in the background and device is passcode locked. It is fixed by checking if we can read or write and if we can't, queue the context and write at the next available time-slot (when the device gets unlocked).
- Note: We do not disable DataProtection or mitigate it in any way. We simply check and wait until we are allowed to read/write before saving the context.
- The other crash happens because we "close" the Profile database via `Profile.shutdown` on enter background and then attempt to write to it while in the background. This throws an sqlite exception.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.


## Test Plan:

<!-- Any useful notes for reviewer explaining how best to test and verify. -->

### Screenshots:

<!-- If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here. -->


## Reviewer Checklist:

- [ ] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [ ] Issues are assigned to at least one epic.
- [ ] Issues include necessary QA labels:
  - [ ] `QA/(Yes|No)`
  - [ ] `release-notes/(include|exclude)`
  - [ ] `bug` / `enhancement`
- [ ] Necessary security reviews have taken place.
- [ ] Adequate test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable)

Brandon-T committed:
Mark Data Contexts as Pending on error.

Brandon-T committed:
Decode base64 key.

Brandon-T committed:
Made init public

iccub commented on Client/Application/Migration.swift:
```diff
@@ -47,6 +47,7 @@ fileprivate extension Preferences {
         migrate(key: "blockPopups", to: Preferences.General.blockPopups)
         migrate(key: "kPrefKeyTabsBarShowPolicy", to: Preferences.General.tabBarVisibility)
         migrate(key: "NightModeStatus", to: Preferences.General.nightMode)
+        migrate(key: "allowBackgroundMediaPlayback", to: Preferences.General.allowBackgroundMediaPlayback)
```
Migration keys were added for preferences in older versions of Brave.
Since this is a new preference, we don't need to migrate it

iccub commented on Client/Frontend/Browser/BackgroundMediaPlayback.swift:
```diff
@@ -0,0 +1,110 @@
+    private var canCleanup = false
+    private var isActive = false
+    private var shutdownRequested = false
+    private weak var profile: Profile?
```
Why do we need to pass `Profile` here? From what I see it's only used for shutdowns

Brandon-T replied:
It's needed because this class handles delayed shutdown when audio is playing. Without it, when we try to `reportSiteVisited`, it will crash due to the `db` being closed. Instead, it checks if the app or audio is active and if it is, it won't shutdown until it becomes inactive or audio stops playing.

danishjafri88 replied:
Can we do it in way such that the handler need not keep reference/ Actually call shutdown?

iccub commented on Client/Frontend/Browser/BackgroundMediaPlayback.swift:
```diff
@@ -0,0 +1,110 @@
+    
+    func userContentController(_ userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage) {
+        if let response = message.body as? String {
+            debugPrint(response)
```
Remove it, log.info is used below

Brandon-T replied:
Fixed.

iccub commented on Client/Frontend/Browser/BraveWebView.swift:
```diff
@@ -19,6 +19,11 @@ class BraveWebView: WKWebView {
     
     init(frame: CGRect, configuration: WKWebViewConfiguration = WKWebViewConfiguration(), isPrivate: Bool = true) {
+        
+        if let data = Data(base64Encoded: "YWx3YXlzUnVuc0F0Rm9yZWdyb3VuZFByaW9yaXR5"), let key = String(data: data, encoding: .utf8) {
```
what is this `YWx3YXlzUnVuc0F0Rm9yZWdyb3VuZFByaW9yaXR5` for?

Brandon-T replied:
It's a field called `alwaysRunsAtForegroundPriority` in `WKWebViewConfiguration` for whether or not a WKWebView can load a request in the background. It's also used in our Unit Tests. It's not exposed so I hid it in base64.

kylehickinson replied:
So this is private WKWebViewConfiguration API? I'd say in general its not great to have an base64 encoded/obfuscated string just lying around in our codebase 😅 Why not obfuscate the old way through String formatting? Either that or definitely needs a comment above 😜

Brandon-T replied:
@kylehickinson The comment is actually right below this conversation lol.

iccub reviewed (changes requested):
Few changes and questions, I have to run this code to do some tests now

Brandon-T committed:
Removed migration and debugPrint.

Brandon-T committed:
Major Cleanup. Added preferences to allow toggling dynamically the background media playback without reloading scripts or reloading the tab.

Brandon-T committed:
Fixed invalid variable name in JS.
Fixed crash in WKWebView observer when the tab dies and the observer receives a delayed event due to a massive lag (such as a PDF or Video loading)

Brandon-T committed:
Fixing Media Background Playback.
Fixing Crash when writing to CoreData in a protected state.
